### PR TITLE
feat: add event & status for ApisixRoute v2

### PR DESCRIPTION
**Pull request by gxthrj**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### CODE REVIEW

File: pkg/ingress/controller.go [hunk 1]
```diff
@@ -44,6 +44,17 @@ import (
 	apisixv1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
 )
 
+const (
+	// SuccessSynced is used when a resource is synced successfully
+	SuccessSynced = "SyncSuccessfully"
+	// MessageResourceSynced is used to specify controller
+	MessageResourceSynced = "%s synced successfully"
+	// FailedSynced is used when a resource synced failed
+	FailedSynced = "SyncFailed"
```
[the remark this answers is not in the document]
Comment (tokers) [reply pkg/ingress/controller.go:53]: PS: we obey the uber go style guide.

File: pkg/kube/apisix/apis/config/v2alpha1/types.go [hunk 2]
```diff
@@ -61,11 +62,18 @@ const (
...
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
 // ApisixRoute is used to define the route rules and upstreams for Apache APISIX.
 type ApisixRoute struct {
 	metav1.TypeMeta   `json:",inline" yaml:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
 	Spec              *ApisixRouteSpec `json:"spec,omitempty" yaml:"spec,omitempty"`
+	Status            ApisixStatus     `json:"status,omitempty" yaml:"status,omitempty"`
+}
```
Comment (tokers) [inline pkg/kube/apisix/apis/config/v2alpha1/types.go:72]: `ApisixTls` and `ApisixUpstream` should also have this field.

File: pkg/ingress/apisix_tls.go [hunk 3]
```diff
@@ -117,6 +123,8 @@ func (c *apisixTlsController) sync(ctx context.Context, ev *types.Event) error {
 			zap.Error(err),
 			zap.Any("ApisixTls", tls),
 		)
+		message := fmt.Sprintf(MessageResourceFailed, TlsController, err.Error())
+		c.recorder.Event(tls, corev1.EventTypeWarning, FailedSynced, message)
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_tls.go]: Will need to query resource objects repeatedly if in `c.handleSyncErr `, which is processed in method sync()

File: pkg/ingress/apisix_upstream.go [hunk 4]
```diff
@@ -122,6 +128,8 @@ func (c *apisixUpstreamController) sync(ctx context.Context, ev *types.Event) er
 	svc, err := c.controller.svcLister.Services(namespace).Get(name)
 	if err != nil {
 		log.Errorf("failed to get service %s: %s", key, err)
+		message := fmt.Sprintf(MessageResourceFailed, UpstreamController, err.Error())
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_upstream.go]: Will need to query resource objects repeatedly if in `c.handleSyncErr `, which is processed in method sync()

File: pkg/ingress/controller.go [hunk 5]
```diff
@@ -44,6 +44,17 @@ import (
 	apisixv1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
 )
 
+const (
+	// SuccessSynced is used when a resource is synced successfully
+	SuccessSynced = "SyncSuccessfully"
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/controller.go]: done

File: pkg/ingress/apisix_route.go [hunk 6]
```diff
@@ -198,14 +210,50 @@ func (c *apisixRouteController) sync(ctx context.Context, ev *types.Event) error
...
+		Type: "APISIXSynced",
+	}
 	if err == nil {
+		message := fmt.Sprintf(MessageResourceSynced, RouteController)
+		if route.GroupVersion() == kube.ApisixRouteV1 {
+			c.recorder.Event(route.V1(), v1.EventTypeNormal, SuccessSynced, message)
+		} else if route.GroupVersion() == kube.ApisixRouteV2alpha1 {
+			c.recorder.Event(route.V2alpha1(), v1.EventTypeNormal, SuccessSynced, message)
+			// build condition
+			condition.Reason = "SyncSuccessfully"
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_route.go]: done

File: pkg/ingress/apisix_route.go [hunk 7]
```diff
@@ -198,14 +210,50 @@ func (c *apisixRouteController) sync(ctx context.Context, ev *types.Event) error
...
 		zap.Error(err),
 	)
+	message := fmt.Sprintf(MessageResourceFailed, RouteController, err.Error())
+	if route.GroupVersion() == kube.ApisixRouteV1 {
+		c.recorder.Event(route.V1(), v1.EventTypeWarning, FailedSynced, message)
+	} else if route.GroupVersion() == kube.ApisixRouteV2alpha1 {
+		c.recorder.Event(route.V2alpha1(), v1.EventTypeWarning, FailedSynced, message)
+		// build condition
+		condition.Reason = "SyncFailed"
+		condition.Status = "False"
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_route.go]: done

File: pkg/ingress/apisix_route.go [hunk 8]
```diff
@@ -198,14 +210,50 @@ func (c *apisixRouteController) sync(ctx context.Context, ev *types.Event) error
 }
 
 func (c *apisixRouteController) handleSyncErr(obj interface{}, err error) {
+	event := obj.(*types.Event)
+	route := event.Object.(kube.ApisixRouteEvent).OldObject
+	// conditions
+	condition := metav1.Condition{
+		Type: "APISIXSynced",
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_route.go]: done

File: pkg/ingress/controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/controller.go]: done

File: pkg/kube/apisix/apis/config/v2alpha1/types.go (hunk 2 above)
Comment (gxthrj) [reply pkg/kube/apisix/apis/config/v2alpha1/types.go:70]: `ApisixTls ` and `ApisixUpstream` will be add in next PR, which is in version v1.

File: pkg/ingress/apisix_upstream.go [hunk 9]
```diff
@@ -122,6 +133,8 @@ func (c *apisixUpstreamController) sync(ctx context.Context, ev *types.Event) er
 	svc, err := c.controller.svcLister.Services(namespace).Get(name)
 	if err != nil {
 		log.Errorf("failed to get service %s: %s", key, err)
+		message := fmt.Sprintf(_messageResourceSyncAborted, _upstreamController, err.Error())
```
Comment (tokers) [inline pkg/ingress/apisix_upstream.go:136]: Ditto for ApisixUpstream.

File: pkg/kube/apisix/apis/config/v2alpha1/types.go (hunk 2 above)
Comment (tokers) [reply pkg/kube/apisix/apis/config/v2alpha1/types.go:70]: OK.

File: pkg/ingress/secret.go [hunk 10]
```diff
@@ -184,9 +184,9 @@ func (c *secretController) onAdd(obj interface{}) {
 		return
 	}
 
-	log.Debugw("secret add event arrived",
-		zap.Any("object", obj),
-	)
+	//log.Debugw("secret add event arrived",
+	//	zap.Any("object", obj),
```
Comment (tokers) [inline pkg/ingress/secret.go:188]: It seems that `zap.Any` outputs the `cert` and `pkey` in `[]bytes` and it's not human-friendly. I think we can just print the `name` and `namespace` of this secret, and do not print the body.

Comment (gxthrj) [reply pkg/ingress/secret.go:188]: Yes, can not print.

File: pkg/ingress/apisix_tls.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tokers) [reply pkg/ingress/apisix_tls.go]: OK.

File: pkg/ingress/apisix_route.go [hunk 11]
```diff
@@ -198,14 +212,50 @@ func (c *apisixRouteController) sync(ctx context.Context, ev *types.Event) error
 }
 
 func (c *apisixRouteController) handleSyncErr(obj interface{}, err error) {
+	event := obj.(*types.Event)
+	route := event.Object.(kube.ApisixRouteEvent).OldObject
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_route.go:216]: Yes, will modify this.

**Comment by gxthrj** [comment]
`kubectl get ar -n test httpbin-route -o yaml`
image

**Comment by gxthrj** [comment]
`kubectl describe ar -n test httpbin-route`
image

File: pkg/ingress/apisix_route.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_route.go]: done

File: pkg/ingress/apisix_route.go [hunk 12]
```diff
@@ -16,30 +16,44 @@ package ingress
...
 	"k8s.io/client-go/util/workqueue"
 
 	"github.com/apache/apisix-ingress-controller/pkg/kube"
+	"github.com/apache/apisix-ingress-controller/pkg/kube/apisix/client/clientset/versioned/typed/config/v2alpha1"
 	"github.com/apache/apisix-ingress-controller/pkg/log"
 	"github.com/apache/apisix-ingress-controller/pkg/types"
 	apisixv1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
 )
 
+const _routeController = "RouteController"
```
Comment (tokers) [inline pkg/ingress/apisix_route.go:40]: I am wondering whether we need a finer granularity component in apisix-ingress-controller. The pod events reporters are "kubelet", "scheduler", it's a whole program, not a module inside it.

File: pkg/ingress/endpoint.go [hunk 13]
```diff
@@ -181,8 +181,8 @@ func (c *endpointsController) onAdd(obj interface{}) {
 	if !c.controller.namespaceWatching(key) {
 		return
 	}
-	log.Debugw("endpoints add event arrived",
-		zap.Any("object", obj))
+	//log.Debugw("endpoints add event arrived",
```
Comment (tokers) [inline pkg/ingress/endpoint.go:184]: Why comment it?

File: pkg/ingress/apisix_upstream.go [hunk 14]
```diff
@@ -122,6 +133,8 @@ func (c *apisixUpstreamController) sync(ctx context.Context, ev *types.Event) er
 	svc, err := c.controller.svcLister.Services(namespace).Get(name)
 	if err != nil {
 		log.Errorf("failed to get service %s: %s", key, err)
+		message := fmt.Sprintf(_messageResourceFailed, _upstreamController, err.Error())
```
Comment (tokers) [inline pkg/ingress/apisix_upstream.go:136]: Can these two lines be encapsulated into one function? they used in many places, and once you need to change, you have to change everywhere.

**Comment by tokers** [comment]
@gxthrj The e2e CI failed.

File: pkg/ingress/endpoint.go (hunk 13 above)
Comment (gxthrj) [reply pkg/ingress/endpoint.go:184]: got error when logging.

**Comment by gxthrj** [comment]
> @gxthrj The e2e CI failed.

It is strange, scale endpoint getting error.

File: pkg/ingress/apisix_route.go (hunk 12 above)
Comment (gxthrj) [reply pkg/ingress/apisix_route.go:40]: Do you mean using `ApisixIngress` is better?

File: pkg/ingress/apisix_upstream.go (hunk 14 above)
Comment (gxthrj) [reply pkg/ingress/apisix_upstream.go:136]: Yes, I will refactor it.

**Comment by gxthrj** [comment]
related to : #394 #395

File: pkg/ingress/endpoint.go (hunk 13 above)
Comment (tokers) [reply pkg/ingress/endpoint.go:184]: error?

File: pkg/ingress/status.go [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/apache/apisix-ingress-controller/pkg/kube"
+	configv2alpha1 "github.com/apache/apisix-ingress-controller/pkg/kube/apisix/apis/config/v2alpha1"
+	"github.com/apache/apisix-ingress-controller/pkg/kube/apisix/client/clientset/versioned/typed/config/v2alpha1"
+)
+
+const (
+	ConditionType        = "ResourcesAvailable"
```
Comment (tokers) [inline pkg/ingress/status.go:31]: Should always set constants unexportable if they are not used by external packages.

File: pkg/ingress/apisix_route.go [hunk 16]
```diff
@@ -197,15 +208,57 @@ func (c *apisixRouteController) sync(ctx context.Context, ev *types.Event) error
...
 }
 
-func (c *apisixRouteController) handleSyncErr(obj interface{}, err error) {
-	if err == nil {
+func (c *apisixRouteController) handleSyncErr(obj interface{}, errOrigin error) {
+	ev := obj.(*types.Event)
+	event := ev.Object.(kube.ApisixRouteEvent)
+	namespace, name, errLocal := cache.SplitMetaNamespaceKey(event.Key)
+	if errLocal != nil {
+		log.Errorf("invalid resource key: %s", event.Key)
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_route.go:216]: done

File: pkg/ingress/apisix_route.go [hunk 17]
```diff
@@ -197,15 +208,57 @@ func (c *apisixRouteController) sync(ctx context.Context, ev *types.Event) error
...
+			if errLocal == nil {
+				message := fmt.Sprintf(_messageResourceSynced, _component)
+				if ar.GroupVersion() == kube.ApisixRouteV1 {
+					c.recorder.Event(ar.V1(), v1.EventTypeNormal, _resourceSynced, message)
+				} else if ar.GroupVersion() == kube.ApisixRouteV2alpha1 {
+					c.recorder.Event(ar.V2alpha1(), v1.EventTypeNormal, _resourceSynced, message)
+					RecordRouteStatus(ar.V2alpha1(), _resourceSynced, CommonSuccessMessage, metav1.ConditionTrue)
+				}
+			} else {
+				log.Errorf("invalid resource key: %s", event.Key)
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_route.go:235]: done

File: pkg/ingress/apisix_route.go [hunk 18]
```diff
@@ -197,15 +208,57 @@ func (c *apisixRouteController) sync(ctx context.Context, ev *types.Event) error
...
 		c.workqueue.Forget(obj)
 		return
 	}
 	log.Warnw("sync ApisixRoute failed, will retry",
 		zap.Any("object", obj),
-		zap.Error(err),
+		zap.Error(errOrigin),
 	)
+	var ar kube.ApisixRoute
+	if event.GroupVersion == kube.ApisixRouteV1 {
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/apisix_route.go]: done

File: pkg/ingress/status.go (hunk 15 above)
Comment (gxthrj) [reply pkg/ingress/status.go:31]: done

File: pkg/ingress/status.go [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/apache/apisix-ingress-controller/pkg/kube/apisix/client/clientset/versioned/typed/config/v2alpha1"
+)
+
+const (
+	_onditionType         = "ResourcesAvailable"
+	_commonSuccessMessage = "Sync Successfully"
+)
+
+// RecordRouteStatus record ApisixRoute v2alpha1 status
+func RecordRouteStatus(ar *configv2alpha1.ApisixRoute, reason, message string, status v1.ConditionStatus) {
```
Comment (tokers) [inline pkg/ingress/status.go:36]: Where is this function used?

Also, if it won't be used by other packages, we don't need to export it.

Comment (gxthrj) [reply pkg/ingress/status.go]: ok, done

File: pkg/ingress/endpoint.go (hunk 13 above)
Comment (tokers) [reply pkg/ingress/endpoint.go:184]: Ping @gxthrj .

File: pkg/ingress/secret.go (hunk 10 above)
Comment (gxthrj) [reply pkg/ingress/secret.go:188]: if log "object" with `zap.Any`, got error ` bad access: nil dereference`.
I changed logging field.

File: pkg/ingress/endpoint.go (hunk 13 above)
Comment (gxthrj) [reply pkg/ingress/endpoint.go:184]: if log "object" with `zap.Any`, got error ` bad access: nil dereference`.
I changed logging field.

File: pkg/ingress/status.go [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+	condition.Reason = reason
+	condition.Status = status
+	condition.Message = message
+	// set to status
+	if ar.Status.Conditions == nil {
+		conditions := make([]metav1.Condition, 0)
+		ar.Status.Conditions = &conditions
+	}
+	meta.SetStatusCondition(ar.Status.Conditions, condition)
+	_, _ = v2alpha1.New(kube.GetApisixClient().ApisixV2alpha1().RESTClient()).ApisixRoutes(ar.Namespace).
```
Comment (tokers) [inline pkg/ingress/status.go:50]: We don't have to new a client each time we want to update the status.

Comment (gxthrj) [reply pkg/ingress/status.go]: done

File: pkg/ingress/status.go [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+
+// recordRouteStatus record ApisixRoute v2alpha1 status
+func recordRouteStatus(ar *configv2alpha1.ApisixRoute, reason, message string, status v1.ConditionStatus) {
+	condition := metav1.Condition{
+		Type: _onditionType,
+	}
+	// build condition
+	condition.Reason = reason
+	condition.Status = status
+	condition.Message = message
```
Comment (tokers) [inline pkg/ingress/status.go:42]:
```suggestion
	condition := metav1.Condition{
		Type: _onditionType,
		Reason: reason,
		Status: status,
		Message: message,
	}
```

Comment (gxthrj) [reply pkg/ingress/status.go]: done

**Review by tokers — APPROVED** [review]

**Pull request merged**